Sources/GeneratorEngine/Engine.swift

```diff
@@ -0,0 +1,81 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import class AsyncHTTPClient.HTTPClient
+@_exported import Crypto
+import struct Logging.Logger
+@_exported import struct SystemPackage.FilePath
+
+/// Cacheable computations engine. Currently the engine makes an assumption that computations produce same results for
+/// the same query values and write results to a single file path.
+public actor Engine {
+  private(set) var cacheHits = 0
+  private(set) var cacheMisses = 0
+
+  public let fileSystem: any FileSystem
+  public let httpClient: HTTPClient
+  public let logger: Logger
+  private let resultsCache: SQLiteBackedCache
+
+  /// Creates a new instance of the ``Engine`` actor.
+  /// - Parameter fileSystem: Implementation of a file system this engine should use.
+  /// - Parameter cacheLocation: Location of cache storage used by the engine.
+  /// - Parameter httpClient: HTTP client to use in queries that need it.
+  /// - Parameter logger: Logger to use during queries execution.
+  public init(
+    _ fileSystem: any FileSystem,
+    _ httpClient: HTTPClient,
+    _ logger: Logger,
+    cacheLocation: SQLite.Location
+  ) {
+    self.fileSystem = fileSystem
+    self.httpClient = httpClient
+    self.logger = logger
+    self.resultsCache = SQLiteBackedCache(tableName: "cache_table", location: cacheLocation, logger)
+  }
+
+  deinit {
+    try! resultsCache.close()
+  }
+
+  /// Executes a given query if no cached result of it is available. Otherwise fetches the result from engine's cache.
+  /// - Parameter query: A query value to execute.
+  /// - Returns: A file path to query's result recorded in a file.
+  public subscript(_ query: some QueryProtocol) -> FilePath {
+    get async throws {
+      var hashFunction = SHA512()
+      query.hash(with: &hashFunction)
+      let key = hashFunction.finalize().description
```

Comment:
`.description`'s output is usually not API. Is this truly the right thing to do?

Reply:
I've changed `SQLiteBackedCache` type to take `some Sequence<UInt8>` as keys to improved the situation. But there are still types like `FileCacheRecord` that are encoded into JSON and we need to serialize hashes as strings for that. But that stems from the fact that `SQLiteBackedCache` is a NoSQL store at this stage. I'll normalize it to make it properly relational in a future PR.